### ISSUE TITLE
Improper ctypes being generated in macro expansion

### ISSUE THREAD

**Issue by adetaylor** [dtolnay/cxx#744]
Do this:

1. Check out this branch with a simplified demo example: https://github.com/dtolnay/cxx/compare/master...adetaylor:improper-ctypes-problem
2. `cd demo; cargo build`. It should _compile_ OK (but not link, simply because I haven't altered the C++ side of the demo example to match)
3. `cargo expand`, which gives the following code for the `#[cxx::bridge]` section:
```rust
#[deny(improper_ctypes)]
#[allow(non_camel_case_types, non_snake_case, clippy::upper_case_acronyms)]
mod ffi2 {
    pub fn make_string(str: &str) -> ::cxx::UniquePtr<::cxx::CxxString> {
        extern "C" {
            #[link_name = "org$blobstore$cxxbridge1$make_string"]
            fn __make_string(str: ::cxx::private::RustStr) -> *mut ::cxx::CxxString;
        }
        unsafe { ::cxx::UniquePtr::from_raw(__make_string(::cxx::private::RustStr::from(str))) }
    }
}
```
4. Note use of `::cxx::private::RustStr` in the above expanded bindings
5. Copy and paste that code to replace the entire `cxx::bridge` mod
6. `cargo build`
7. It doesn't even compile. It says:
```
   Compiling demo v0.0.0 (/Users/adetaylor/dev/cxx/demo)
error: `extern` block uses type `str`, which is not FFI-safe
  --> demo/src/main.rs:17:35
   |
17 |             fn __make_string(str: ::cxx::private::RustStr) -> *mut ::cxx::CxxString;
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^ not FFI-safe
   |
note: the lint level is defined here
  --> demo/src/main.rs:11:8
   |
11 | #[deny(improper_ctypes)]
   |        ^^^^^^^^^^^^^^^
   = help: consider using `*const u8` and a length instead
   = note: string slices have no C equivalent

error: aborting due to previous error
```

As far as I can see:
* The `#[deny(improper_ctypes)]` lint is ignored in the expansion of the `cxx::bridge` macro...
* ... and the code in `improper.rs` which tries to solve this by using void pointers doesn't seem to be having the desired effect.

Impact: none at all for production use of `cxx`. I'm fiddling around with expanding `cxx::bridge` macros a different way in order to make `rust-analyzer` happy which is how I came across it. Thought I should let you know.

rustc version:
```
$ rustc --version
rustc 1.50.0 (cb75ad5db 2021-02-10)
```

**Comment by dtolnay** [comment]
I would accept a PR to fix this. It would also be worth filing a rustc bug with a minimal repro. I don't understand why the lint would trigger in one case but not the other. We should fix the macro regardless, because presumably a future version of rustc will fix the bug and we don't want that to start breaking existing users of this crate.

**Comment by dtolnay** [comment]
The cases I was able to find are fixed by #819, #821, #826, #827. I'll go ahead and close. Let's file separate issues if any other specific cases of this are observed.
